core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java

```diff
@@ -418,7 +418,19 @@ public Block copyPositions(int[] positions, int offset, int length)
             }
             newIds[i] = newId;
         }
-        return new DictionaryBlock(dictionary.copyPositions(positionsToCopy.elements(), 0, positionsToCopy.size()), newIds);
+        Block compactDictionary = dictionary.copyPositions(positionsToCopy.elements(), 0, positionsToCopy.size());
+        if (positionsToCopy.size() == length) {
+            // discovered that all positions are unique, so return the unwrapped underlying dictionary directly
+            return compactDictionary;
+        }
+        return new DictionaryBlock(
```

Comment:
please add tests to `io.trino.block.TestDictionaryBlock`

Reply:
Added